pymongo/lock.py

```diff
@@ -0,0 +1,126 @@
+import platform
+import threading
+import weakref
+
+
+class _ForkLock:
+    """
+    Represents a lock that is tracked upon instantiation using a WeakSet and
+    reset by pymongo upon forking.
+    """
+
+    # Needs to lock all locks without causing deadlocks elsewhere.
+    # So we wait for all locks to be released, and then lock _global_lock
+    # to prevent more locks while we fork.
+
+    # Issues
+    # - Cannot cause deadlocks by grabbing a lock when there are nested
+    # locks.
+    # - Needs to grab all locks before another thread can.
+    # - Needs to deal with fact only calling thread will be alive in
+    # child process.
+    # - Needs to have minimal performance impact.
+
+    # Solution
+    # When forking
+    # - Lower a Condition indicating no other locks should be acquired.
+    # Other locks may not acquire **unless 1+ is already alive.**
+    # - Upon 0 locks being held, lock all of them.
+    # - Fork
+    # - Unlock all locks.
+    # - Release Condition and notify_all.
+    # When locking
+    # - Is Condition lowered and are no locks active? => Wait and block
+    # - Is Condition lowered and are locks active? => Continue (likely a
+    # dependent lock)
+    # - Is Condition not lowered and are locks active => Continue (not
+    # forking)
+    # - Atomic increment lock counter. (done first and last to ensure
+    # correct state)
+    # - Acquire lock.
+    # - Critical region (and any nested critical regions)
+    # - Release lock.
+    # - Decrement lock counter.
+
+    if platform.python_implementation() == "CPython":
```

Comment:
We should be able to remove this CPython specific code here after fixing the deadlock issues.